### PR TITLE
Ds4 26 assembly line accept recipes

### PR DESCRIPTION
**Pull request by Siole**
Code is not testet yet but needs feedback that this is actually what we want the method to be doing.

### CODE REVIEW

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 1]
```diff
@@ -91,9 +116,12 @@ public Item takeOutputItem() {
         throw new UnsupportedOperationException("Not implemented yet");
     }
 
-    /*private boolean hasItemsForRecipe(Item[] items) {
-        throw new UnsupportedOperationException("Not implemented yet");
-        FIXME: Not used yet
-    }*/
+    private boolean hasItemsForRecipe(Item[] items) {
+        StartAssemblyRequest startAssemblyRequest = new StartAssemblyRequest();
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:120]: the recipe to fetch should be passed as a parameter

File: backend/CommonAssembly/src/main/java/dk/sdu/controllab/commonassembly/AssemblyService.java [hunk 2]
```diff
@@ -16,7 +18,7 @@ public interface AssemblyService {
      * precondition: API is connected to the assembly station, and the assembly station is in IDLE state
      * postcondition: the assembly station is in EXECUTING state
      */
-    void startAssembly();
+    void startAssembly(UUID id);
```
Comment (MagnusHJensen) [inline backend/CommonAssembly/src/main/java/dk/sdu/controllab/commonassembly/AssemblyService.java:21]:
```suggestion
    void startAssembly(UUID recipeId);
```

File: backend/Assembly/src/test/java/dk/sdu/controllab/assembly/service/AssemblyServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.UUID;
+
+import static org.assertj.core.api.Assertions.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class AssemblyServiceImplTest {
```
Comment (MagnusHJensen) [inline backend/Assembly/src/test/java/dk/sdu/controllab/assembly/service/AssemblyServiceImplTest.java:16]: I'm not sure this brings too much value :/

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 4]
```diff
@@ -66,18 +75,34 @@ public GetStatusResponse getStatus() {
...
-        process.addProperty("ProcessId", id);
-        String processId = process.toString();
-
-        try {
-            MqttMessage message = new MqttMessage(processId.getBytes());
-            message.setQos(0);
-            mqttClient.publish("emulator/operation", message);
-        } catch (MqttException e) {
-            e.printStackTrace();
+    public void startAssembly(UUID id) {
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:78]:
```suggestion
    public void startAssembly(UUID recipeId) {
```

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 5]
```diff
@@ -66,18 +75,34 @@ public GetStatusResponse getStatus() {
...
+            for (int i = 0; i < assemblyInventory.getAmount(); i++) {
+                itemList.add(assemblyInventory.getItem());
+            }
+        }
+        Item[] items = itemList.toArray(new Item[0]);
+
+        if (!hasItemsForRecipe(items)) {
+            throw new IllegalArgumentException("Provided items do not match the recipe");
+        } else{
+            startAssemblyRequest.getRecipeId(); //the machine cannot do anything with this, but now it has the recipe
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:93]: This should also remove the used items from the db table.

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 6]
```diff
@@ -91,9 +116,12 @@ public Item takeOutputItem() {
...
-    /*private boolean hasItemsForRecipe(Item[] items) {
-        throw new UnsupportedOperationException("Not implemented yet");
-        FIXME: Not used yet
-    }*/
+    private boolean hasItemsForRecipe(Item[] items) {
+        StartAssemblyRequest startAssemblyRequest = new StartAssemblyRequest();
+        Arrays.sort(items);
+        List<Item> recipeItems = recipeService.getRecipe(startAssemblyRequest.getRecipeId()).getInputs();
+        Arrays.sort(recipeItems.toArray());
+        return Arrays.equals(items, recipeItems.toArray());
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:124]:
```suggestion
        Item[] recipeItems = recipeService.getRecipe(startAssemblyRequest.getRecipeId()).getInputs().toArray();
        Arrays.sort(recipeItems);
        return Arrays.equals(items, recipeItems);
```

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 7]
```diff
@@ -91,9 +116,12 @@ public Item takeOutputItem() {
...
     }
 
-    /*private boolean hasItemsForRecipe(Item[] items) {
-        throw new UnsupportedOperationException("Not implemented yet");
-        FIXME: Not used yet
-    }*/
+    private boolean hasItemsForRecipe(Item[] items) {
+        StartAssemblyRequest startAssemblyRequest = new StartAssemblyRequest();
+        Arrays.sort(items);
+        List<Item> recipeItems = recipeService.getRecipe(startAssemblyRequest.getRecipeId()).getInputs();
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:122]: Recipe could be null, leading to an NPE.

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java [hunk 8]
```diff
@@ -66,18 +75,34 @@ public GetStatusResponse getStatus() {
...
+        StartAssemblyRequest startAssemblyRequest = new StartAssemblyRequest();
+        List<AssemblyInventoryEntry> assemblyInventoryEntries = assemblyRepository.findAll();
+        List<Item> itemList = new ArrayList<>();
+
+        for(AssemblyInventoryEntry assemblyInventory : assemblyInventoryEntries){
+            for (int i = 0; i < assemblyInventory.getAmount(); i++) {
+                itemList.add(assemblyInventory.getItem());
+            }
+        }
+        Item[] items = itemList.toArray(new Item[0]);
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:88]:
```suggestion
        Item[] items = itemList.toArray();
```

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/controller/requests/StartAssemblyRequest.java [hunk 9]
```diff
@@ -2,7 +2,10 @@
 
 import lombok.Getter;
 
+import java.util.UUID;
+
 @Getter
 public class StartAssemblyRequest {
-    private int recipeId;
+    //int according to class diagram but UUID needed for getting a recipe
```
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/controller/requests/StartAssemblyRequest.java:9]:
```suggestion
```

File: backend/Assembly/pom.xml [hunk 10]
```diff
@@ -58,6 +58,19 @@
...
+            <artifactId>CommonRecipe</artifactId>
+            <version>0.0.1-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <version>3.23.1</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (MagnusHJensen) [inline backend/Assembly/pom.xml:72]: This should not be needed.
`assertThat` has the same assertions, and is a part of `spring-boot-starter-test`

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java (hunk 5 above)
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:93]:
```suggestion
```

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java (hunk 1 above)
Comment (MagnusHJensen) [inline backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java:120]:
```suggestion
```

File: backend/Assembly/src/test/java/dk/sdu/controllab/assembly/service/AssemblyServiceImplTest.java (hunk 3 above)
Comment (Siole) [reply backend/Assembly/src/test/java/dk/sdu/controllab/assembly/service/AssemblyServiceImplTest.java:16]: Agreed

File: backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java (hunk 7 above)
Comment (Siole) [reply backend/Assembly/src/main/java/dk/sdu/controllab/assembly/service/AssemblyServiceImpl.java]: NPE?

**Review by MagnusHJensen — APPROVED** [review]

**Pull request merged**